File: Homework1/Hw1/Parser.cs
namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args,
        out double val1,
        out CalculatorOperation operation,
        out double val2)
    {
        if (!IsArgLengthSupported(args) || !double.TryParse(args[0], out val1) || !double.TryParse(args[2], out val2))
        {
            throw new ArgumentException("Invalid arguments for calculating");
        }
        operation = ParseOperation(args[1]);
        if (operation == CalculatorOperation.Undefined)
        {
            throw new InvalidOperationException("Invalid operation");
        }
    }

    private static bool IsArgLengthSupported(string[] args) => args.Length == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        switch (arg)
        {
            case "+":
                return CalculatorOperation.Plus;
            case "-":
                return CalculatorOperation.Minus;
            case "*":
                return CalculatorOperation.Multiply;
            case "/":
                return CalculatorOperation.Divide;
            default:
                return CalculatorOperation.Undefined;
        }
    }
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    args.Length = 3

let parseOperation (arg : string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | arg -> CalculatorOperation.Undefined
    
let parseCalcArguments(args : string[]) =
    let isSuccessVal1, val1 = Double.TryParse(args[0])
    let isSuccessVal2, val2 = Double.TryParse(args[2])
    let operation = parseOperation(args[1])
    if not (isSuccessVal1 && isSuccessVal2 && isArgLengthSupported(args)) then invalidArg "" ""
    elif operation = CalculatorOperation.Undefined then invalidOp ""
    else {arg1 =  val1; arg2 = val2; operation = operation}